src/array/cuda/labor_sampling.cu

```diff
@@ -0,0 +1,807 @@
+    const auto w2 = B[in_idx];
+    const auto data = data_arr ? data_arr[in_idx] : in_idx;
+    return thrust::make_tuple(in_row, row, u, data, w / min((FloatType)1, c * w2 * ps));
+  }
+};
+
+template <typename FloatType>
+struct StencilOp {
+  const FloatType * cs;
+  template <typename IdType>
+  __host__ __device__
+  auto operator() (IdType in_row, FloatType ps, FloatType rnd) {
+    return rnd <= cs[in_row] * ps;
+  }
+};
+
+template <typename IdType, typename FloatType, typename ps_t, typename A_t>
+struct StencilOpFused {
+  const uint64_t rand_seed;
+  const IdType * idx_coo;
+  const FloatType * cs;
+  const ps_t probs;
+  const A_t A;
+  const IdType * subindptr;
+  const IdType * rows;
+  const IdType * indptr;
+  const IdType * indices;
+  const IdType * nids;
+  __device__
+  auto operator() (IdType idx) {
+    const auto in_row = idx_coo[idx];
+    const auto ps = probs[idx];
+    IdType rofs = idx - subindptr[in_row];
+    const IdType row = rows[in_row];
+    const auto in_idx = indptr[row] + rofs;
+    const auto u = indices[in_idx];
+    const auto t = nids ? nids[u] : u;  // t in the paper
+    curandStatePhilox4_32_10_t rng;
+    if (labor)
+      // rolled random number r_t is a function of the random_seed and t
+      curand_init(123123, rand_seed, t, &rng);
+    else  // reduces to neighbor sampling, see r_t vs r_{ts} in arXiv:2210.13339
+      curand_init(rand_seed, idx, 0, &rng);
+    const float rnd = curand_uniform(&rng);
+    return rnd <= cs[in_row] * A[in_idx] * ps;
+  }
+};
+
+template <typename IdType, typename FloatType>
+struct TransformOpMean {
+  const IdType * ds;
+  const FloatType * ws;
+  __host__ __device__
+  auto operator() (IdType idx, FloatType ps) {
+    return ps * ds[idx] / ws[idx];
+  }
+};
+
+struct TransformOpMinWith1 {
+  template <typename FloatType>
+  __host__ __device__
+  auto operator() (FloatType x) {
+    return min((FloatType)1, x);
+  }
+};
+
+template <typename IdType>
+struct IndptrFunc {
+  const IdType * indptr;
+  __host__ __device__
+  auto operator() (IdType row) {
+    return indptr[row];
+  }
+};
+
+template <typename FloatType>
+struct SquareFunc {
+  __host__ __device__
+  auto operator() (FloatType x) {
+    return thrust::make_tuple(x, x * x);
+  }
+};
+
+struct TupleSum {
+    template <typename T>
+    __host__ __device__
+    T operator()(const T &a, const T &b) const {
+        return thrust::make_tuple(thrust::get<0>(a) + thrust::get<0>(b),
+            thrust::get<1>(a) + thrust::get<1>(b));
+    }
+};
+
+template <typename IdType, typename FloatType>
+struct DegreeFunc {
+  const IdType num_picks;
+  const IdType * rows;
+  const IdType * indptr;
+  const FloatType * ds;
+  IdType * in_deg;
+  FloatType * cs;
+  __host__ __device__
+  auto operator() (IdType tIdx) {
+    const auto out_row = rows[tIdx];
+    const auto d = indptr[out_row + 1] - indptr[out_row];
+    in_deg[tIdx] = d;
+    cs[tIdx] = num_picks / (ds ? ds[tIdx] : (FloatType)d);
```

Comment:
Is `ds` a necessary argument for this operator?  If `ds` is not `nullptr`, it means the graph is weighted, and `cs` needs to be recomputed anyway.  Did I miss something?

Reply:
I see.